Feature/pm 421 temperature and ipd to robot monitor

Joris-JW opened the pull request:
Closes PM-421

## Description
This PR integrates a first version of the RQT robot monitor package which is dedicated to show the analytics of the MARCH exoskeleton during runtime. This version provides the joint temperatures and validates if the input device is connected and publishing at the desired frequency.

New diagnostics can easily be added to the main updater if the given message is in the right format. Using the analyzer.yaml the diagnostics are logically grouped so the robot monitor can categorize the data. More information about the structure and usage of the diagnostics and monitor can be found at the official ROS website; http://wiki.ros.org/diagnostics.

## Changes
* Created a new package for the robot monitor
* Integrated the temperature check and IPD connection check
* Created a .yaml file to order the diagnostics for the robot monitor 

## Example (and how to test)
In order to test the current setup the right data must be provided, this can be accomplished by playing a random .bag file originating from the MARCH exoskeleton. Use the ```roslaunch march_rqt_robot_monitor march_rqt_robot_monitor.launch``` file to execute and start the monitor.

![image](https://user-images.githubusercontent.com/45165354/80595277-bf3b1f00-8a24-11ea-8747-7b0078b5d3f3.png)

Joris-JW committed:
Created a new catkin package for the robot monitor related files.

Including the right packages and libraries in the CMakeLists.txt and package.xml.

Joris-JW committed:
Created two analysers, one for the temperature measurements and the other for monitoring the frequency of a topic.

The temperature measurements use the threshold limits defined in the march safety yaml file. The frequency check is used to verify if the alive topic of the input device is publishing at the right frequency.

Joris-JW committed:
Added the general updater with contains the temperature and the IPD check. Also created a script which is launched when the robot monitor is launched to make sure the updater is alive.

The analyzers.yaml categorizes the information of the diagnostic topic to the right format for the robot monitor.

Joris-JW committed:
Merge branch 'develop' of https://github.com/project-march/monitor into feature/PM-421-temperature-and-ipd-to-robot-monitor

Joris-JW committed:
Fixed package dependencies.

Joris-JW committed:
Removed redundant spaces.

RutgerVanBeek commented on march_rqt_robot_monitor/src/march_rqt_robot_monitor/diagnostic_analyzers/temperature.py:
```diff
@@ -0,0 +1,85 @@
+            stat.add(' Topic error ', 'No events recorded.')
+            return stat
+
+        if self._temperature < LOWER_THRESHOLD_VALID_VALUE or self._temperature < UPPER_THRESHOLD_VALID_VALUE:
+            stat.summary(DiagnosticStatus.WARN, 'No valid temperature value.')
+            return stat
```
I indeed got a 'No valid temperature value.' for all the joint when launching the simulation. 
```suggestion
        if self._temperature < LOWER_THRESHOLD_VALID_VALUE or self._temperature > UPPER_THRESHOLD_VALID_VALUE:
            stat.summary(DiagnosticStatus.WARN, 'No valid temperature value.')
            return stat
```

RutgerVanBeek commented on march_rqt_robot_monitor/src/march_rqt_robot_monitor/diagnostic_analyzers/temperature.py:
```diff
@@ -0,0 +1,85 @@
+from sensor_msgs.msg import Temperature
+
+
+LOWER_THRESHOLD_VALID_VALUE = 0
+UPPER_THRESHOLD_VALID_VALUE = 100
```
```suggestion
LOWER_THRESHOLD_VALID_VALUE_TEMPERATURE = 0
UPPER_THRESHOLD_VALID_VALUE_TEMPERATURE = 100
```

RutgerVanBeek commented on march_rqt_robot_monitor/src/march_rqt_robot_monitor/diagnostic_analyzers/temperature.py:
```diff
@@ -0,0 +1,85 @@
+        if self._temperature < self._thresholds_warning:
+            stat.summary(DiagnosticStatus.OK, 'OK')
+        elif self._thresholds_warning < self._temperature < self._thresholds_non_fatal:
+            stat.summary(DiagnosticStatus.WARN, 'Temperature in warning zone.')
```
```suggestion
            stat.summary(DiagnosticStatus.WARN, 'Temperature almost to high.')
```

RutgerVanBeek commented on march_rqt_robot_monitor/src/march_rqt_robot_monitor/updater.py:
```diff
@@ -0,0 +1,23 @@
+    rospy.init_node('Diagnostic_updater')
+
+    updater = diagnostic_updater.Updater()
+    updater.setHardwareID('MARCH IV')
```
;)
```suggestion
    updater.setHardwareID('MARCH IVc')
```

Olavhaasie replied:
You might want to change this to a parameter

Joris-JW replied:
nice.

RutgerVanBeek reviewed (changes requested):
Looks good overall! Some minor suggestions and I have two small extra comments. 

May be you can include the actual temperature and not only whether it is too high or almost too high etc..

I disconnect the Input Device connected and got something along the way of 'frequency too low' . I think you can just make that 'input device disconnected'.

Olavhaasie commented on march_rqt_robot_monitor/launch/march_rqt_robot_monitor.launch:
```diff
@@ -0,0 +1,15 @@
+        <rosparam command="load" file="$(find march_rqt_robot_monitor)/resource/analyzers.yaml"/>
+    </node>
+
+    <node pkg="rqt_robot_monitor" type="rqt_robot_monitor" name="rqt_robot_monitor">
+    </node>
```
```suggestion
    <node pkg="rqt_robot_monitor" type="rqt_robot_monitor" name="rqt_robot_monitor"/>
```

Olavhaasie commented on march_rqt_robot_monitor/launch/march_rqt_robot_monitor.launch:
```diff
@@ -0,0 +1,15 @@
+<launch>
+    <node name="march_rqt_robot_monitor_node"
+          pkg="march_rqt_robot_monitor"
+          type="march_rqt_robot_monitor_node"
+          output="screen"
+          required="true" >
```
[`required`](https://wiki.ros.org/roslaunch/XML/node) means that when this node crashes every other node should be shut down. I don't think this is necessarily the case here. The monitor is not necessary for the robot to be able to run. You could add `respawn`, which will cause roslaunch to respawn the node if it crashes.
```suggestion
    <node name="march_rqt_robot_monitor_node"
          pkg="march_rqt_robot_monitor"
          type="march_rqt_robot_monitor_node"
          output="screen"
          respawn="true">
```

Joris-JW replied:
nice addition

Olavhaasie commented on march_rqt_robot_monitor/src/march_rqt_robot_monitor/diagnostic_analyzers/temperature.py:
```diff
@@ -0,0 +1,85 @@
+    def __init__(self, joint_name, topic):
+        rospy.Subscriber(topic, Temperature, self.cb)
+
+        self._default_thresholds = \
+            rospy.get_param('/safety_node/default_temperature_threshold', None)
+        self._thresholds_warning = \
+            rospy.get_param('/safety_node/temperature_thresholds_warning/{jn}'.format(jn=joint_name), None)
+        self._thresholds_non_fatal = \
+            rospy.get_param('/safety_node/temperature_thresholds_non_fatal/{jn}'.format(jn=joint_name), None)
+        self._thresholds_fatal = \
+            rospy.get_param('/safety_node/temperature_thresholds_fatal/{jn}'.format(jn=joint_name), None)
```
I think for this path to always work you have to remove the preceding '/'. That means it will always look in the root of the parameter server. However, the march nodes, when launched together, will be placed in the `march` namespace. So in order to always look relative to this node's current namespace you must remove the `/`.
```suggestion
        self._default_thresholds = \
            rospy.get_param('/safety_node/default_temperature_threshold', None)
        self._thresholds_warning = \
            rospy.get_param('safety_node/temperature_thresholds_warning/{jn}'.format(jn=joint_name), None)
        self._thresholds_non_fatal = \
            rospy.get_param('safety_node/temperature_thresholds_non_fatal/{jn}'.format(jn=joint_name), None)
        self._thresholds_fatal = \
            rospy.get_param('safety_node/temperature_thresholds_fatal/{jn}'.format(jn=joint_name), None)
```

Olavhaasie reviewed (changes requested):
The `analyzers.yaml` is contained in the `resource` directory. However, normally we would put config files (yaml) inside the `config` directory. `resource` is mostly used for images or .ui files.

Finally, I don't know if this is possible, but every input device could be displayed separately. Then you know which one is lost. However, for this to work you would need something to identify them with in the alive message, which does not currently exist. So, perhaps something for another time.

Joris-JW committed:
Fixed comments on PR.

Changed temperature threshold naming, moved the analyzers.yaml to the config folder, used relative parameter naming in the temperature.py constructor. Also replaced the required tag in the launch file with a respawn tag.

Joris-JW commented:
> Finally, I don't know if this is possible, but every input device could be displayed separately. Then you know which one is lost. However, for this to work you would need something to identify them with in the alive message, which does not currently exist. So, perhaps something for another time.

This seems like a nice touch! However, not sure how to implement this . Maybe, now that the package exists, we can discus on how we want to implement new features.

Joris-JW committed:
Fixed list DIRECTORY should be sorted.

RutgerVanBeek commented on march_rqt_robot_monitor/src/march_rqt_robot_monitor/diagnostic_analyzers/temperature.py:
```diff
@@ -50,7 +51,7 @@ def diagnostics(self, stat):
             stat.summary(DiagnosticStatus.OK, 'OK')
         elif self._thresholds_warning < self._temperature < self._thresholds_non_fatal:
-            stat.summary(DiagnosticStatus.WARN, 'Temperature in warning zone.')
+            stat.summary(DiagnosticStatus.WARN, 'Temperature almost to high.')
```
```suggestion
            stat.summary(DiagnosticStatus.WARN, 'Temperature almost too high.')
```

RutgerVanBeek reviewed (changes requested):
Some small comments left. 

Also from my previous review, is it possible to add the actual temperature to this message and not just at which level.

Joris-JW committed:
Update march_rqt_robot_monitor/src/march_rqt_robot_monitor/diagnostic_analyzers/temperature.py

Co-authored-by: RutgerVanBeek <51922475+RutgerVanBeek@users.noreply.github.com>
